### ISSUE TITLE
How to calculate indicators such as PSNR?

### ISSUE THREAD

**Issue by Wen-Hui-Ma** [PeizhiYan/gaussian-dejavu#9]

**Comment by PeizhiYan** [comment]
Hi!

I'm not sure if you're asking about general methods for evaluating image reconstruction quality, such as SSIM and PSNR, or how to assess our model's reconstruction quality using these metrics. Could you please clarify?

**Comment by PeizhiYan** [comment]
Hi! Thanks for your interest in our work.

This repo is mainly for demonstration purpose, it is a cleaned version and does not contain all the code we wrote. 

I can send you the code segments we used for computing L1, LPIPS, SSIM and PSNR:

```[python]
from lpips import LPIPS # lpips loss

############################
## Evaluation Metrics
############################

def compute_l1(img_gt, img_overlay):
    # compute the mean L1 distance between a pair of images
    return np.mean(3*abs(img_gt - img_overlay)) # x3 because we only want to average across each channel


class LPIPS_LOSS_VGG():
    def __init__(self, device) -> None:
        # image should be RGB, IMPORTANT: normalized to [-1,1]
        self.lpips_loss_fn_vgg = LPIPS(net='vgg').to(device) # closer to "traditional" perceptual loss, when used for optimization

    def compute(self, preds, targs):
        """
        input size NCHW
        input RGB range [-1, 1]
        """
        loss = self.lpips_loss_fn_vgg(preds, targs)
        loss = torch.mean(loss) # average on batch
        return loss
    

lpips_loss_vgg = LPIPS_LOSS_VGG(device)

def compute_lpips(img_gt, img_overlay):
    # compute the LPIPS loss
    
    gt = torch.from_numpy(img_gt * 2 - 1).to(device)[None].permute(0,3,1,2)
    over = torch.from_numpy(img_overlay * 2 - 1).to(device)[None].permute(0,3,1,2)
    
    return float(lpips_loss_vgg.compute(gt, over).item())

from skimage.metrics import structural_similarity as ssim
from skimage.metrics import peak_signal_noise_ratio as psnr

def rgb_to_grayscale(img):
    # Ensure the image has three dimensions [H, W, C]
    if img.ndim != 3 or img.shape[2] != 3:
        raise ValueError("Input image must be [H, W, C] with three channels.")
    # Convert image to grayscale
    gray = 0.299 * img[:, :, 0] + 0.587 * img[:, :, 1] + 0.114 * img[:, :, 2]
    return gray

def compute_ssim(img_gt, img_overlay):
    return ssim(img_gt, img_overlay, multichannel=True, channel_axis=2, data_range=1)
    #return ssim(rgb_to_grayscale(img_gt), rgb_to_grayscale(img_overlay), data_range=1)

def compute_psnr(img_gt, img_overlay):
    return psnr(img_gt, img_overlay)
```

When computing the loss for the facial region, we first use Face Parsing from this work (https://github.com/zllrunning/face-parsing.PyTorch) to segment the facial region:
```[python]
# Parsing Net
from face_parsing.model import BiSeNet
segnet = BiSeNet(n_classes=19)
segnet.cuda()
segnet.load_state_dict(torch.load('./tools/face_parsing/79999_iter.pth'))
segnet.eval()

def run_face_parsing(img):
    # input standardized to 0 ~ 1
    n_classes = 19
    to_tensor = transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize((0.485, 0.456, 0.406), (0.229, 0.224, 0.225)),
    ])
    with torch.no_grad():
        img = to_tensor(img)
        img = torch.unsqueeze(img, 0)
        img = img.cuda()
        out = segnet(img)[0]
        parsing = out.squeeze(0).cpu().numpy().argmax(0)
    ###################
    ## Release CUDA
    import gc
    gc.collect() # colloct memory
    torch.cuda.empty_cache() # empty cuda
    return parsing


def class_remapping(parsing_mask):
    """
    Old Classes:
        atts = {1: 'skin', 2: 'l_brow', 3: 'r_brow', 4: 'l_eye', 5: 'r_eye', 
                6: 'eye_g', 7: 'l_ear', 8: 'r_ear', 9: 'ear_r', 10: 'nose', 
                11: 'mouth', 12: 'u_lip', 13: 'l_lip', 14: 'neck', 15: 'neck_l', 
                16: 'cloth', 17: 'hair', 18: 'hat'}
    New Classes (11 classes):
            0: background
            1: face skin
            2: eye brows
            3: eyes
            4: nose
            5: upper lip
            6: lower lip
            7: ears
            8: hair
            9: hat
            10: eyeglasses
            11: mouth
    """
    #new_mask = np.copy(silhouette).astype(np.int)
    new_mask = np.zeros(parsing_mask.shape).astype(np.int)
    def process(parsing_mask, old_class, new_class):
        one_mask = np.where(parsing_mask == old_class, 1, 0)
        return one_mask, new_class*one_mask
    mapping = { 1: 1,
                2: 2, 
                3: 2, 
                4: 3, 
                5: 3, 
                10: 4, 
                12: 5, 
                13: 6,
                7: 7,
                8: 7,
                17: 8,
                18: 9,
                6: 10,
                11: 11} # format  old_class: new_class
    for old_class in mapping.keys():
        one_mask, class_mask = process(parsing_mask, old_class=old_class, new_class=mapping[old_class])
        new_mask = new_mask * (1 - one_mask) + class_mask    
    return new_mask


def read_img(img_path, resize=512):
    # 3-channel image
    img = cv2.imread(img_path)
    img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
    img = cv2.resize(img, (resize, resize))
    img = np.array(img, dtype=np.float32) / 255. # 0 ~ 1
    return img


def read_mask(mask_path, resize=512):
    # binary mask 8-bit
    mask = cv2.imread(mask_path, 0) / 255
    mask = cv2.resize(mask, (resize, resize))
    return mask
    
def apply_mask(img, mask):
    # set backgound color to gray
    img_copy = np.copy(img) * 2 - 1
    for c in range(img_copy.shape[2]):
        img_copy[:,:,c] = img_copy[:,:,c] * mask
    return np.clip(img_copy / 2 + 0.5, 0, 1)
```

You can render and save the images, then use the following code to evaluate the rendered image:

```[python]
# evaluation metric functions
eval_functions = {
    'l1': compute_l1,
    'lpips': compute_lpips,
    'ssim': compute_ssim,
    'psnr': compute_psnr,
}

# load images
targ_img = read_img(img_path)                   # this is the ground-truth image
pred_img = read_img(rendered_img_path) # this is the rendered image

# parsing
parsing_mask = run_face_parsing(targ_img)
parsing_mask_ = run_face_parsing(pred_img)

# get face mask
facemask = np.zeros(parsing_mask.shape)
facemask[np.where((parsing_mask >= 1) & (parsing_mask <= 13))] = 1
facemask_ = np.zeros(parsing_mask_.shape)
facemask_[np.where((parsing_mask_ >= 1) & (parsing_mask_ <= 13))] = 1
facemask = facemask * facemask_

# get head mask
headmask_ = np.zeros(parsing_mask_.shape)
headmask_[np.isin(parsing_mask_, [1,2,3,4,5,6,7,8,10,11,12,13,17])] = 1
headmask = np.zeros(parsing_mask.shape)
headmask[np.isin(parsing_mask, [1,2,3,4,5,6,7,8,10,11,12,13,17])] = 1
headmask = headmask * headmask_

# apply head mask
targ_img_masked_head = apply_mask(targ_img, headmask)
pred_img_masked_head = apply_mask(pred_img, headmask)
    
# apply face mask
targ_img_masked_face = apply_mask(targ_img, facemask)
pred_img_masked_face = apply_mask(pred_img, facemask)

# compute the losses
for metric_name in eval_functions.keys():
    results[metric_name]['all'].append(eval_functions[metric_name](targ_img_masked_head, pred_img_masked_head))
    results[metric_name]['facial'].append(eval_functions[metric_name](targ_img_masked_face, pred_img_masked_face))

```

**Comment by PeizhiYan** [comment]
You are welcome!
